### PR TITLE
Source Pinterest: support OAuth

### LINKED ISSUE #7754: Source Pinterest: support oauth
## Tell us about the problem you're trying to solve
With the release of Airbyte Cloud, we need to start supporting Oauth for this connector, since it's the recommended way of authenticating users into a SaaS application.

If this connector doesn't support oauth already (i.e: doesn't accept a client_id and client_secret) then we need to update its spec to accept those parameters. There are two ways to do this:

If the connector already supports some auth mechanism like api_key, I suggest that this be a oneof nested inside a top-level field called "authentication":

{ authentication: { type: object oneOf: [ // api key, // oauth ] } }

If the connector only supports webflow oauth, then no changes are needed to the properties format and we will only need to add annotations.

See the connector spec reference in the docs for reference on how a oneof can be implemented.

This should be done in a backwards compatible manner i.e: users currently supplying authentication info in the config's top-level should not be impacted by this change.
##Acceptance criteria
The connector supports oauth webflow authentication with client_id/client_secret in a backwards compatible manner
Oauth properties are annotated properly.

### PR DESCRIPTION
**Pull request by arsenlosenko**
## What
Resolving:
https://github.com/airbytehq/airbyte/issues/7754

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-pinterest/source_pinterest/source.py [hunk 1]
```diff
@@ -289,6 +289,7 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
 class SourcePinterest(AbstractSource):
     @staticmethod
     def get_authenticator(config):
+        config = config.get("credentials") or config
         user_pass = (config.get("client_id") + ":" + config.get("client_secret")).encode("ascii")
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-pinterest/source_pinterest/source.py:293]: nit: `base64_encoded` is better

File: airbyte-integrations/connectors/source-pinterest/source_pinterest/source.py [hunk 2]
```diff
@@ -289,6 +289,7 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
...
         auth = "Basic " + standard_b64encode(user_pass).decode("ascii")
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-pinterest/source_pinterest/source.py:294]: Make sure `.decode("utf-8")` is not the option to use.

File: docs/integrations/sources/pinterest.md [hunk 3]
```diff
@@ -71,6 +70,7 @@ Boards streams - 10 calls per sec / per user / per app
 
 | Version | Date       | Pull Request                                             | Subject                                           |
 | :------ | :--------- | :------------------------------------------------------- | :------------------------------------------------ |
+| 0.1.3   | 2022-09-02 | [16271](https://github.com/airbytehq/airbyte/pull/16271) | Implement OAuth authentication
```
Comment (bazarnov) [inline docs/integrations/sources/pinterest.md:73]: Added support for `OAuth2.0` authentication method

File: docs/integrations/sources/pinterest.md [hunk 4]
```diff
@@ -14,11 +14,10 @@ Please read [How to get your credentials](https://developers.pinterest.com/docs/
...
 2. In the left navigation bar, click **Sources**. In the top-right corner, click **+new source**.
 3. On the Set up the source page, enter the name for the Pinterest connector and select **Pinterest** from the Source type dropdown. 
-4. Enter your `client_id`
-5. Enter your `client_secret`
-6. Enter your `refresh_token`
-7. Enter the `start_date` you want your sync to start from
-8. Click **Set up source**
+4. Enter the `start_date` you want your sync to start from
+5. Choose `OAuth2.0` in `Authorization Method` list
+6. Proceed with OAuth authentication of your account in the pop-up window
```
Comment (bazarnov) [inline docs/integrations/sources/pinterest.md:19]: There should be a button that says: `Authenticate your account` or similar, please reflect this as a step in public docs.

**Review by bazarnov — CHANGES_REQUESTED** [review]

**Review by lazebnyi — APPROVED** [review]

**Review by bazarnov — APPROVED** [review]

**Comment by arsenlosenko** [comment]
/publish connector=connectors/source-pinterest run-tests=false

**Comment by sherifnada** [comment]
@arsenlosenko why were tests skipped here?

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/PinterestOAuthFlow.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Supplier;
+import org.apache.http.client.utils.URIBuilder;
+
+/**
+ * Following docs from https://developers.pinterest.com/docs/getting-started/authentication
+ */
+public class PinterestOAuthFlow extends BaseOAuth2Flow {
```
Comment (sherifnada) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/PinterestOAuthFlow.java:30]: @arsenlosenko could you add unit tests for this function?

Comment (arsenlosenko) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/PinterestOAuthFlow.java:30]: @sherifnada yes, already working on unit tests for this function in scope of this PR:
https://github.com/airbytehq/airbyte/pull/16393

**Comment by arsenlosenko** [comment]
> @arsenlosenko why were tests skipped here?

@sherifnada the tests were skipped due to low rate limits in Pinterest API (300 requests/day), due to that we couldn't run complete test suite here

**Comment by YowanR** [comment]
We got access to a higher level access for this account. Can we just run the test and make sure we are good now, please?
cc @arsenlosenko @lazebnyi

**Pull request merged**